### PR TITLE
brscan5: 1.2.9-0 -> 1.3.0-0

### PR DESCRIPTION
**Pull request by AkechiShiro**
## Description of changes

Bump version, changelog not available as the driver is proprietary.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Just tagging maintainer @mattchrist for review. 

Missing `nixpkgs-review` and nixos tests, that need to be run I believe.

### CODE REVIEW

**Comment by mattchrist** [comment]
The brscan5 nixos test is now failing:
```
nix-build --no-out-link nixos/tests/brscan5.nix
```
The output of `scanimage -L` doesn't include the model number:
```
device `brother5:net1;dev0' is a Brother  
device `brother5:net1;dev1' is a Brother  
```

The tests expect it should look like:
```
device `brother5:net1;dev0' is a Brother b ADS-1200
device `brother5:net1;dev1' is a Brother a ADS-1200
```

**Comment by mattchrist** [comment]
I wonder if the format of the brsanenetdevice.cfg file has changed in this new version? If so maybe the nixos module needs updating as well.

**Comment by AkechiShiro** [comment]
Got it @mattchrist I will try and take a look at the NixOS module, if I can fix the tests (NixOS tests right ?)

Then I will open a second PR to fix this one hopefully or should I just put everything under this one ?

**Comment by mattchrist** [comment]
Please fix it here, since the package update alone breaks some functionality of the NixOS module.

**Comment by AkechiShiro** [comment]
I may be able to take a look soon, sorry for having taking so long.

File: pkgs/applications/graphics/sane/backends/brscan5/default.nix [hunk 1]
```diff
@@ -10,15 +10,15 @@ let
 in
 stdenv.mkDerivation rec {
   pname = "brscan5";
-  version = "1.2.9-0";
+  version = "1.3.0-0";
   src = {
     "i686-linux" = fetchurl {
       url = "https://download.brother.com/welcome/dlf104034/${pname}-${version}.i386.deb";
-      sha256 = "ac23c9a435818955e7882ab06380adf346203ff4e45f384b40e84b8b29642f07";
+      sha512 = "sha512-G6afTP6pf47FK2Zpl6Gm69W6nQp0JQ+npaJ8sWhQM98xBbN74kb/kCXhZ8tkYixzC+7EQm01R6lNajpmMoTT3w==";
```
Comment (RaitoBezarius) [inline pkgs/applications/graphics/sane/backends/brscan5/default.nix:17]: we have no reason to use sha512 yet.

**Comment by AkechiShiro** [comment]
Hi again @mattchrist, unfortunately I do not have a Brother ADS-1200, can I still fix the tests using any other scanner ?

**Comment by mattchrist** [comment]
That test emulates the network mode for setting up an ADS-1200, you don't actually need to have one (or any scanner at all really) to run and fix the integration test. You can run it yourself from your branch by running `nix-build --no-out-link nixos/tests/brscan5.nix` in the root of your clone of the nixpkgs repository.

The test is defined in this file: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/brscan5.nix

It makes sure that after setting up the pretend scanners via the [`hardware.sane.brscan5.netDevices`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/hardware/sane_extra_backends/brscan5.nix) NixOS option, `scanimage` successfully detects the configured network scanners that were configured by [this package](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/hardware/sane_extra_backends/brscan5_etc_files.nix).

**Comment by mattchrist** [comment]
I'll also note that setup of the brother scanner via brscan5 driver in NixOS is fairly hacky. We're doing some binary patching to look in /etc instead of the hardcoded /opt location (which won't exist on nixos).
https://github.com/NixOS/nixpkgs/blob/64f5298190b799ad6efed5e6b71c71d815005cdf/pkgs/applications/graphics/sane/backends/brscan5/default.nix#L52

I seem to remember checking your PR earlier to make sure this particular offset was the right location to patched and to see if brother added any other hardcoded paths to the file that needed patching (running `strings -n 10 --radix=d opt/brother/scanner/brscan5/libsane-brother5.so.1.0.7` on the so file).
So I don't think there's any new weirdness that needs handling, but wanted to call it out for awareness in case I missed something.

**Comment by AkechiShiro** [comment]
I saw during the test that there is this message just before the failure : 
```bash
machine # [08:15:04.583648] [bjnp] create_broadcast_socket: ERROR - bind socket to local address failed - Cannot assign requested address
````
Could it be because of that, that the `scanimage -L` does not show the ADS-1200 ?

**Comment by mattchrist** [comment]
I think that's unrelated.

I took another look at the new driver, and it seems the offset that needed patching has changed:
```diff
diff --git a/pkgs/applications/graphics/sane/backends/brscan5/default.nix b/pkgs/applications/graphics/sane/backends/brscan5/default.nix
index 93c3ffb1a6c8..391176470a3f 100644
--- a/pkgs/applications/graphics/sane/backends/brscan5/default.nix
+++ b/pkgs/applications/graphics/sane/backends/brscan5/default.nix
@@ -34,8 +34,8 @@ stdenv.mkDerivation rec {
   postPatch =
     let
       patchOffsetBytes =
-        if system == "x86_64-linux" then 84632
-        else if system == "i686-linux" then 77396
+        if system == "x86_64-linux" then 86528
+        else if system == "i686-linux" then 79140
         else throw "Unsupported system: ${system}";
     in
     ''
```
This seems to work for me.

**Comment by AkechiShiro** [comment]
I'm getting this error on my side with the same offsets running the tests : 
> machine # [    7.771191] systemd-udevd[438]: /nix/store/b86g2g3lm4rn3iw680g2s9gq4ngb0lcs-udev-rules/49-brother-mfp-brscan5-1.0.2-2.rules:17 Invalid key 'SYSFS'.

**Comment by AkechiShiro** [comment]
Do you have any pointers to how I could fix the issue because I don't understand why the version change causes the output of `scanimage -L` to change like that. I can't really figure out what is exactly wrong, is there any debugging steps I can do in the NixOS VM (like build and pop a shell into the VM environment to investigate further) ?

**Comment by swflint** [comment]
Is there any movement on this?  It looks like this would fix issues with getting my scanner to work.

**Comment by mattchrist** [comment]
I'm waiting for @AkechiShiro to apply the changes I mentioned in https://github.com/NixOS/nixpkgs/pull/254992#issuecomment-1831903153.
It also appears another reviewer would prefer a sha256 hash instead of the sha512. https://github.com/NixOS/nixpkgs/pull/254992/files#r1405413790

**Comment by AkechiShiro** [comment]
Sorry @mattchrist I didn't get to it yet yet, I will push all requested changes by the end of this week

**Comment by AkechiShiro** [comment]
I confirm that correcting the offset fixes the test : 
```
(finished: must succeed: strace scanimage -L 2>&1, in 16.55 seconds)
machine: must succeed: cat /etc/opt/brother/scanner/brscan5/brsanenetdevice.cfg
(finished: must succeed: cat /etc/opt/brother/scanner/brscan5/brsanenetdevice.cfg, in 0.04 seconds)
machine: must succeed: scanimage -L
machine # [   15.788614] systemd-udevd[485]: /nix/store/zsac4rlb72g3cp66f675qpa8j9w9qnil-udev-rules/49-brother-mfp-brscan5-1.0.2-2.rules:17 Invalid key 'SYSFS'
machine # [   18.970700] systemd-udevd[485]: /nix/store/zsac4rlb72g3cp66f675qpa8j9w9qnil-udev-rules/49-brother-mfp-brscan5-1.0.2-2.rules:17 Invalid key 'SYSFS'
(finished: must succeed: scanimage -L, in 4.15 seconds)
device `brother5:net1;dev0' is a Brother b ADS-1200
device `brother5:net1;dev1' is a Brother a ADS-1200

(finished: run the VM test script, in 20.76 seconds)
test script finished in 20.82s
cleanup
kill machine (pid 6)
machine # qemu-kvm: terminating on signal 15 from pid 4 (/nix/store/vagb0sjv83ybi435i6iiv10hjrdghph9-python3-3.10.12/bin/python3.10)
(finished: cleanup, in 0.20 seconds)
kill vlan (pid 5)
/nix/store/x48cvgmb6w9naj91wsv2vki9hapd6zvw-vm-test-run-brscan5
```

Will push the changes.

**Comment by AkechiShiro** [comment]
Should I rebase on master since it's been awhile before marking this PR ready for review @mattchrist ?

**Comment by mattchrist** [comment]
Great! I don't think a rebase is needed since there's no merge conflicts.

I'll give it one more look after you mark the PR ready and add my review.

**Review by mattchrist — APPROVED** [review]
Confirmed nixos test passes, and confirmed this package still works for scanning with my DS-640.

**Comment by AkechiShiro** [comment]
@swflint do you think this update should be backported as your scanner might not work without it ? It could also be alright for it to be in unstable, just wanted your opinion on this.

**Comment by swflint** [comment]
I work off of unstable as it is.  That being said, it should probably be backported for the sake of other users.

**Review by mattchrist — APPROVED** [review]

**Pull request merged**